File: plugins/rider/src/main/kotlin/com/shortsleevestudio/gamescript/handlers/DialogHandlers.kt
package com.shortsleevestudio.gamescript.handlers

import com.google.gson.JsonObject
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileChooser.FileChooser
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory
import com.intellij.openapi.fileChooser.FileSaverDescriptor
import com.intellij.openapi.fileChooser.FileChooserFactory
import com.intellij.openapi.ui.Messages

/**
 * Handlers for native dialog operations.
 * Implements: dialog:openSqlite, dialog:saveSqlite, dialog:openCsv,
 * dialog:saveCsv, dialog:selectFolder
 */
class DialogHandlers(private val context: HandlerContext) {

    /**
     * Get all handlers as a map.
     */
    fun getHandlers(): Map<String, suspend (JsonObject) -> Unit> = mapOf(
        "dialog:openSqlite" to ::handleOpenSqlite,
        "dialog:saveSqlite" to ::handleSaveSqlite,
        "dialog:openCsv" to ::handleOpenCsv,
        "dialog:saveCsv" to ::handleSaveCsv,
        "dialog:selectFolder" to ::handleSelectFolder
    )

    private suspend fun handleOpenSqlite(message: JsonObject) {
        val id = message.get("id")?.asString ?: return

        ApplicationManager.getApplication().invokeLater {
            try {
                val descriptor = FileChooserDescriptorFactory.createSingleFileDescriptor()
                    .withFileFilter { vf ->
                        vf.extension?.lowercase() in listOf("db", "sqlite", "sqlite3")
                    }
                    .withTitle("Open SQLite Database")
                    .withDescription("Select a SQLite database file")

                val chosen = FileChooser.chooseFile(descriptor, context.project, null)

                if (chosen != null) {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to false, "filePath" to chosen.path)
                    )
                } else {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to true)
                    )
                }
            } catch (e: Exception) {
                LOG.error("Error showing file chooser", e)
                context.postResponse(id, "dialog:result", false, error = e.message)
            }
        }
    }

    private suspend fun handleSaveSqlite(message: JsonObject) {
        val id = message.get("id")?.asString ?: return
        val defaultName = message.get("defaultName")?.asString ?: "database.db"

        ApplicationManager.getApplication().invokeLater {
            try {
                val descriptor = FileSaverDescriptor(
                    "Save SQLite Database",
                    "Choose where to save the database",
                    "db", "sqlite", "sqlite3"
                )

                val dialog = FileChooserFactory.getInstance().createSaveFileDialog(descriptor, context.project)
                val baseDir = context.getWorkspaceFolder()
                val result = dialog.save(baseDir, defaultName)

                if (result != null) {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to false, "filePath" to result.file.path)
                    )
                } else {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to true)
                    )
                }
            } catch (e: Exception) {
                LOG.error("Error showing save dialog", e)
                context.postResponse(id, "dialog:result", false, error = e.message)
            }
        }
    }

    private suspend fun handleOpenCsv(message: JsonObject) {
        val id = message.get("id")?.asString ?: return

        ApplicationManager.getApplication().invokeLater {
            try {
                val descriptor = FileChooserDescriptorFactory.createSingleFileDescriptor()
                    .withFileFilter { vf ->
                        vf.extension?.lowercase() == "csv"
                    }
                    .withTitle("Open CSV File")
                    .withDescription("Select a CSV file to import")

                val chosen = FileChooser.chooseFile(descriptor, context.project, null)

                if (chosen != null) {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to false, "filePath" to chosen.path)
                    )
                } else {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to true)
                    )
                }
            } catch (e: Exception) {
                LOG.error("Error showing CSV file chooser", e)
                context.postResponse(id, "dialog:result", false, error = e.message)
            }
        }
    }

    private suspend fun handleSaveCsv(message: JsonObject) {
        val id = message.get("id")?.asString ?: return
        val defaultName = message.get("defaultName")?.asString ?: "export.csv"

        ApplicationManager.getApplication().invokeLater {
            try {
                val descriptor = FileSaverDescriptor(
                    "Save CSV File",
                    "Choose where to save the CSV file",
                    "csv"
                )

                val dialog = FileChooserFactory.getInstance().createSaveFileDialog(descriptor, context.project)
                val baseDir = context.getWorkspaceFolder()
                val result = dialog.save(baseDir, defaultName)

                if (result != null) {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to false, "filePath" to result.file.path)
                    )
                } else {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to true)
                    )
                }
            } catch (e: Exception) {
                LOG.error("Error showing CSV save dialog", e)
                context.postResponse(id, "dialog:result", false, error = e.message)
            }
        }
    }

    private suspend fun handleSelectFolder(message: JsonObject) {
        val id = message.get("id")?.asString ?: return
        val title = message.get("title")?.asString ?: "Select Folder"

        ApplicationManager.getApplication().invokeLater {
            try {
                val descriptor = FileChooserDescriptorFactory.createSingleFolderDescriptor()
                    .withTitle(title)
                    .withDescription("Select a folder")

                val chosen = FileChooser.chooseFile(descriptor, context.project, context.getWorkspaceFolder())

                if (chosen != null) {
                    // Convert to relative path if within workspace
                    val workspacePath = context.getWorkspacePath()
                    val chosenPath = chosen.path

                    val relativePath = if (workspacePath != null && chosenPath.startsWith(workspacePath)) {
                        chosenPath.removePrefix(workspacePath).removePrefix("/").ifEmpty { "." }
                    } else {
                        null
                    }

                    if (relativePath != null) {
                        context.postResponse(
                            id, "dialog:result", true,
                            mapOf("cancelled" to false, "filePath" to relativePath)
                        )
                    } else {
                        // Folder outside workspace - show error
                        Messages.showErrorDialog(
                            context.project,
                            "Selected folder must be within the workspace.",
                            "Invalid Selection"
                        )
                        context.postResponse(
                            id, "dialog:result", true,
                            mapOf("cancelled" to true)
                        )
                    }
                } else {
                    context.postResponse(
                        id, "dialog:result", true,
                        mapOf("cancelled" to true)
                    )
                }
            } catch (e: Exception) {
                LOG.error("Error showing folder chooser", e)
                context.postResponse(id, "dialog:result", false, error = e.message)
            }
        }
    }

    companion object {
        private val LOG = Logger.getInstance(DialogHandlers::class.java)
    }
}

File: plugins/dotnet/visual-studio/GameScript.VisualStudio/Handlers/DialogHandlers.cs
using System.Text.Json;
using Microsoft.VisualStudio.Extensibility.Shell;

namespace GameScript.VisualStudio;

/// <summary>
/// Handlers for native dialog operations.
/// Implements: dialog:openSqlite, dialog:saveSqlite, dialog:openCsv,
/// dialog:saveCsv, dialog:selectFolder
/// </summary>
public static class DialogHandlers
{
    public static void Register(MessageMediator mediator, HandlerContext context)
    {
        mediator.Register("dialog:openSqlite", (msg, ct) => HandleOpenSqlite(msg, context, ct));
        mediator.Register("dialog:saveSqlite", (msg, ct) => HandleSaveSqlite(msg, context, ct));
        mediator.Register("dialog:openCsv", (msg, ct) => HandleOpenCsv(msg, context, ct));
        mediator.Register("dialog:saveCsv", (msg, ct) => HandleSaveCsv(msg, context, ct));
        mediator.Register("dialog:selectFolder", (msg, ct) => HandleSelectFolder(msg, context, ct));
    }

    private static async Task HandleOpenSqlite(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        string id = message.GetRequiredString("id");

        OpenFileDialogOptions options = new()
        {
            Title = "Open SQLite Database",
            Filters =
            {
                new FileDialogFilter("SQLite Database", ".db", ".sqlite", ".sqlite3"),
                new FileDialogFilter("All Files", ".*")
            }
        };

        await ShowOpenFileDialogAsync(id, options, context, ct);
    }

    private static async Task HandleSaveSqlite(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        string id = message.GetRequiredString("id");
        string defaultName = message.GetOptionalString("defaultName") ?? "database.db";

        SaveAsFileDialogOptions options = new()
        {
            Title = "Save SQLite Database",
            DefaultFileName = defaultName,
            Filters =
            {
                new FileDialogFilter("SQLite Database", ".db", ".sqlite", ".sqlite3"),
                new FileDialogFilter("All Files", ".*")
            }
        };

        await ShowSaveFileDialogAsync(id, options, context, ct);
    }

    private static async Task HandleOpenCsv(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        string id = message.GetRequiredString("id");

        OpenFileDialogOptions options = new()
        {
            Title = "Open CSV File",
            Filters =
            {
                new FileDialogFilter("CSV Files", ".csv"),
                new FileDialogFilter("All Files", ".*")
            }
        };

        await ShowOpenFileDialogAsync(id, options, context, ct);
    }

    private static async Task HandleSaveCsv(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        string id = message.GetRequiredString("id");
        string defaultName = message.GetOptionalString("defaultName") ?? "export.csv";

        SaveAsFileDialogOptions options = new()
        {
            Title = "Save CSV File",
            DefaultFileName = defaultName,
            Filters =
            {
                new FileDialogFilter("CSV Files", ".csv"),
                new FileDialogFilter("All Files", ".*")
            }
        };

        await ShowSaveFileDialogAsync(id, options, context, ct);
    }

    private static async Task HandleSelectFolder(JsonElement message, HandlerContext context, CancellationToken ct)
    {
        string id = message.GetRequiredString("id");
        string title = message.GetOptionalString("title") ?? "Select Folder";

        try
        {
            OpenFolderDialogOptions options = new()
            {
                Title = title
            };

            Uri? folderUri = await context.Extensibility.Shell().ShowOpenFolderDialogAsync(options, ct);

            if (folderUri != null)
            {
                string chosenPath = folderUri.LocalPath;
                string? workspacePath = await context.GetWorkspacePathAsync(ct);

                // Convert to relative path if within workspace
                string? relativePath = null;
                if (!string.IsNullOrEmpty(workspacePath) && chosenPath.StartsWith(workspacePath, StringComparison.OrdinalIgnoreCase))
                {
                    relativePath = chosenPath.Substring(workspacePath.Length).TrimStart(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
                    if (string.IsNullOrEmpty(relativePath))
                    {
                        relativePath = ".";
                    }
                }

                if (relativePath != null)
                {
                    await SendDialogResultAsync(id, false, relativePath, context, ct);
                }
                else
                {
                    // Folder outside workspace - show error via prompt
                    await context.Extensibility.Shell().ShowPromptAsync(
                        "Selected folder must be within the workspace.",
                        PromptOptions.OK,
                        ct);

                    await SendDialogCancelledAsync(id, context, ct);
                }
            }
            else
            {
                await SendDialogCancelledAsync(id, context, ct);
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[DialogHandlers] Error showing folder dialog: {ex.Message}");
            await context.PostResponseAsync(id, "dialog:result", false, error: ex.Message, cancellationToken: ct);
        }
    }

    /// <summary>
    /// Helper to show an open file dialog and send the result.
    /// </summary>
    private static async Task ShowOpenFileDialogAsync(
        string id,
        OpenFileDialogOptions options,
        HandlerContext context,
        CancellationToken ct)
    {
        try
        {
            Uri? fileUri = await context.Extensibility.Shell().ShowOpenFileDialogAsync(options, ct);

            if (fileUri != null)
            {
                await SendDialogResultAsync(id, false, fileUri.LocalPath, context, ct);
            }
            else
            {
                await SendDialogCancelledAsync(id, context, ct);
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[DialogHandlers] Error showing open dialog: {ex.Message}");
            await context.PostResponseAsync(id, "dialog:result", false, error: ex.Message, cancellationToken: ct);
        }
    }

    /// <summary>
    /// Helper to show a save file dialog and send the result.
    /// </summary>
    private static async Task ShowSaveFileDialogAsync(
        string id,
        SaveAsFileDialogOptions options,
        HandlerContext context,
        CancellationToken ct)
    {
        try
        {
            Uri? fileUri = await context.Extensibility.Shell().ShowSaveAsFileDialogAsync(options, ct);

            if (fileUri != null)
            {
                await SendDialogResultAsync(id, false, fileUri.LocalPath, context, ct);
            }
            else
            {
                await SendDialogCancelledAsync(id, context, ct);
            }
        }
        catch (Exception ex)
        {
            System.Diagnostics.Debug.WriteLine($"[DialogHandlers] Error showing save dialog: {ex.Message}");
            await context.PostResponseAsync(id, "dialog:result", false, error: ex.Message, cancellationToken: ct);
        }
    }

    /// <summary>
    /// Send a successful dialog result with a file path.
    /// </summary>
    private static Task SendDialogResultAsync(
        string id,
        bool cancelled,
        string filePath,
        HandlerContext context,
        CancellationToken ct)
    {
        return context.PostResponseAsync(id, "dialog:result", true, new Dictionary<string, object?>
        {
            ["cancelled"] = cancelled,
            ["filePath"] = filePath
        }, cancellationToken: ct);
    }

    /// <summary>
    /// Send a cancelled dialog result.
    /// </summary>
    private static Task SendDialogCancelledAsync(
        string id,
        HandlerContext context,
        CancellationToken ct)
    {
        return context.PostResponseAsync(id, "dialog:result", true, new Dictionary<string, object?>
        {
            ["cancelled"] = true
        }, cancellationToken: ct);
    }
}
